Added unit tests and Github work flows to run tests and lints

Mart-Bogdan opened the pull request:
Lints would post all clippy warnings as annotation to pool request

Mart-Bogdan committed:
Added unit tests and Github work flows to run tests and lints

Mart-Bogdan commented:
By the way. It should add tests count as comment to PR, like here: https://github.com/Mart-Bogdan/dirstat-rs/pull/3#issuecomment-1355791986

I guess it is because it is not merged yet. Or because I am not a maintainer. 

![image](https://user-images.githubusercontent.com/775038/208209546-dcfd6800-1d8c-4773-9883-4f71728e1d8b.png)

Mart-Bogdan commented:
> By the way. It should add tests count as comment to PR
> I guess it is because it is not merged yet. Or because I am not a maintainer.


Oh. That's unfortuante:
> [Publish Unit Tests Results](https://github.com/scullionw/dirstat-rs/actions/runs/3717361316/jobs/6304655757#step:4:22)
This action is running on a pull_request event for a fork repository. It cannot do anything useful like creating check runs or pull request comments. To run the action on fork repository pull requests, see https://github.com/EnricoMi/publish-unit-test-result-action/blob/v1.20/README.md#support-fork-repositories-and-dependabot-branches

Mart-Bogdan committed:
remove deprecated github actions

scullionw commented on .github/workflows/check-and-lint.yaml:
```diff
@@ -30,52 +25,44 @@ jobs:
-          token: ${{ secrets.GITHUB_TOKEN }}
-          args: --all-features
-          name: Clippy Output
+# TODO clippy action is deprecated. Perhaps use https://github.com/actions-rs/clippy-check but it require more work
```
what do you think about adding TODO's to issues instead to not clutter the code?

Mart-Bogdan replied:
Yeah. Seems reasonable.

When I first added this action it worked. But then lot's changed around and now this action don't work at all.

And another one displays annotations in github UI only in private PR's.

I guess it's reasonable to create issue for fix of clippy warnings. And perhaps add simple check that would just print clippy errors to console, but as other PR. I'll convert this to issue.

scullionw commented on .github/workflows/test.yaml:
```diff
@@ -64,6 +64,9 @@ jobs:
           path: artifacts
 
       - name: Publish Unit Test Results
+        # TODO this won't work for PRs from contributors,
```
if it doesn't work for all PRs i wouldn't add this, what do you think?

Mart-Bogdan replied:
Yeah, problem is that it works only fir PRs created by repo owner for security reasons. It worked in my local repository.

But actually author of this action has workaround. It's just not so clearly documented.

Idea is to run tests in context of PR, than upload test results as file, and hava another action triggered by this event, that receives test results as json/xml and then generates message inside PR.

I guess this code is not needed in PR. 


I guess cargo suity won't be neded as well. It is required to generate test result in JUnit format.

Strange that test results parser is default functionality on gitLAB but is hidden behind restricted actions on GH

Commit:
Addressed PR comments. Removed not working CI steps.

Mart-Bogdan commented:
Hi, @scullionw.
I've addressed commets, and removed dead code in actions.

Would there be other comments regarding tests?


Sorry for slow reaction time. I have issues with internet and electricity, and also some chores.

Mart-Bogdan commented:
I actually don't like to put actual files into repo, but I wasn't able to come with anything more clever. 


Perhaps generating them as random files with predefined seed, hm.

Random files are mostly needed to work with compressed files.

Mart-Bogdan commented:
Hello, @scullionw. Happy holidays. Happy new year. 

So what about this PR?)

scullionw commented on Cargo.toml:
```diff
@@ -19,12 +19,19 @@ atty = "0.2.14"
+const_format = "0.2.23"
+# TODO update to 0.16 when it is released. 
+# It fixes https://github.com/la10736/rstest/issues/158, which is not showtopper, but minor annoyance.
+rstest = "0.15.0"
```
looks like 0.16 is released

Mart-Bogdan replied:
nice.))

scullionw commented on src/tests.rs:
```diff
@@ -0,0 +1,225 @@
+//! Be aware that tests are run in parallel. For this reason we must be sure to use
```
since we are testing the public api, can we move the tests outside src/ to be an integration test? ie tests/pathlength.rs tests/sizes.rs etc?

see https://doc.rust-lang.org/book/ch11-03-test-organization.html#integration-tests

Mart-Bogdan replied:
Need to take look at your link. And about pros/cons


For example in branch for windows, i also used private function call in tests: `ffi::set_file_compression(file.path(), compress)?;`

https://github.com/Mart-Bogdan/dirstat-rs/blob/max-path/src/ffi.rs#L126-L169

https://github.com/Mart-Bogdan/dirstat-rs/blob/max-path/src/tests.rs#L248


But generally, looks like you are correct. This are actually integration tests. And they interract with external world.

So looks good. Shall try doing it.

Shall `test-data` dir go under tests dir as well?

scullionw commented on src/tests.rs:
```diff
@@ -0,0 +1,225 @@
+
+use crate::{DiskItem, FileInfo};
+// warn: don't remove `as &str` after macro invocation.
+// It breaks type checker in Intellij Rust IDE
```
i'd prefer not adding IDE specific notes to the source

Mart-Bogdan replied:
well. Perhaps this one is not needed. I think `as &str` thing is already fixed upstream in `const_format` crate. I shall remove this

scullionw commented on src/tests.rs:
```diff
@@ -0,0 +1,225 @@
+99999999/999999/99999999/99999999999/99999999/999999999/9999999999/";
+
+const PATH_1_FULL: &str = concatcp!(TEST_DATA_DIR, LONG_PATH_DIR, PATH_1) as &str;
+//noinspection SpellCheckingInspection
```
same, I think you can disable spellcheck in your idea instead

Mart-Bogdan replied:
You aren't using spellchecker plugin in vscode? Or whichever IDE you are using.

I think disabling spellchecker altogether is harmful.

It's useful in other places. And I sometimes have problems with spelling :-(

![image](https://user-images.githubusercontent.com/775038/217520197-a0204a51-2975-4c78-b4a6-e77c2484ab75.png)

I don't think it's such a big deal, having this comment. We can add second one for VSCode spell checker as well :-)

scullionw commented on src/tests.rs:
```diff
@@ -0,0 +1,225 @@
+    let result = FileInfo::from_path(test_path, true);
+
+    // Then
+    if let Result::Ok(FileInfo::Directory { volume_id }) = result {
```
this is equivalent to `expect("Can not get file info)",

also "Ok" is already in scope, no need for "Result" :)

Mart-Bogdan replied:
Yeah, you are right about Result. 


About expect, just tried it. Don't work :-(   it only covers `Ok` but don't allow to match Directory

the error is:
```
  --> src/tests.rs:63:9
   |
63 |     let FileInfo::Directory { volume_id } = result.expect("Can not get file info");
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ pattern `File { .. }` not covered
   |
   = note: `let` bindings require an "irrefutable pattern", like a `struct` or an `enum` with only one variant
```

So anyway `if let` would be needed.

scullionw commented on src/tests.rs:
```diff
@@ -0,0 +1,225 @@
+fn test_file_size_8KiB() {
+    const DIR: &str = concatcp!(TEST_DATA_DIR, "test_file_size/") as &str;
+    // do not rename it into `_` it would cause immediate destrucion after creation
+    let _guard = CleanUpGuard { path: DIR };
```
this is an anti-pattern IMO. if we want to rely on drop to cleanup, we should need to use the type itself. (like create_dir taking a &GuardedPath as argument or something). easy to forget otherwise!

Mart-Bogdan replied:
I've eused pattern mentioned here: https://stackoverflow.com/questions/38253321/what-is-a-good-way-of-cleaning-up-after-a-unit-test-in-rust

Unfortunately I do not know better way of doing something after code is executed. Regardles if panic was raised.

Somebody are proposing setting panic hook, but that looks like dirty hack.

```
std::panic::set_hook(Box::new(|_| {
    cleanup();
```

What do you mean create_dir should take Guard as argument? Won't this cause it to clean-up straight after function call? 


But ok. I can think on it after other items are resolved. Perhaps you would be able to come with something better for clean-up.

Mart-Bogdan replied:
I'll move all the code into integration tests first, I think.

scullionw reviewed (approved):
cool, thanks! lgtm once we address a few things

scullionw commented:
do all tests pass without the other PR? if not lets mark them as #[should_panic], and then remove them in the other PR, so that each PR by itself is mergeable

Mart-Bogdan commented:
> do all tests pass without the other PR? if not lets mark them as #[should_panic], and then remove them in the other PR, so that each PR by itself is mergeable

I did not include tests that won't work without other PR. Besides, they won't compile. They are still in separate branch.

Mart-Bogdan commented:
Ah. shuch tests currently are marked as `#[cfg(not(windows))]`  hmm.

Commit:
Resolve some of PR comments. Not all yet.

Mart-Bogdan commented:
Hello, @scullionw, regarding integration tests. 

I'm thinking. This crate exports library as well as executable. Is it required, and do we need to keep backward compatibility for lib?
